### PR TITLE
Add strace equivalent on ABI boundary.

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: mathetake <takeshi@tetrate.io>

ref: https://github.com/envoyproxy/envoy/pull/14107#issuecomment-739711288

### CODE REVIEW

**Comment by mathetake** [comment]
looks a little bit tricky for WAVM...

File: src/v8/v8.cc [hunk 1]
```diff
@@ -554,13 +585,19 @@ void V8::registerHostFunctionImpl(std::string_view module_name, std::string_view
       store_.get(), type.get(),
       [](void *data, const wasm::Val params[], wasm::Val[]) -> wasm::own<wasm::Trap> {
         auto func_data = reinterpret_cast<FuncData *>(data);
+        func_data->vm_->integration()->trace(
+            "[vm->host] " + func_data->name_ + "(" +
+            printValues(params, std::tuple_size<std::tuple<Args...>>::value) + ")");
```
Comment (PiotrSikora) [inline src/v8/v8.cc:590]: Is there any reason why this isn't using `printValues(params, sizeof...(Args)))` like the original code?

Comment (mathetake) [reply src/v8/v8.cc:590]: no reason. just I didn't see the detail of the original impl. That looks much better, thanks!

Comment (PiotrSikora) [reply src/v8/v8.cc:590]: It's linked in the comment that you referenced here :P

**Comment by PiotrSikora** [comment]
> looks a little bit tricky for WAVM...

Since we're using V8 by default, this doesn't have to be a blocker and we can always add it later (if you prefer it).

**Comment by mathetake** [comment]
not sure ETA for WAVM since it looks like I need to be familiar with WAVM's api so let's come back to WAVM later. I will create the issue after merging.

**Review by PiotrSikora — COMMENTED** [review]
Thanks, the code looks good now.

One thing that we need to consider, though, is that now those strings will be always evaluated, and calls to Envoy will be always made, even when the log level is other than `trace`, which results in wasted compute.

`ENVOY_LOG(trace, ...)` macro, on the other hand, only evaluates string when the log level is `trace`.

We might need to get the log level updates from Envoy, or at least a getter, and only process the string when it's at the `trace` level... or do you think that's premature optimization?

**Comment by mathetake** [comment]
> One thing that we need to consider, though, is that now those strings will be always evaluated, and calls to Envoy will be always made, even when the log level is other than trace, which results in wasted compute.

yeah, that's exactly what I am thinking about. strace should be triggered by "userland" so how about having ABI to enable/disable strace in host?

**Comment by mathetake** [comment]
ah, that's problematic since we cannot observe `_start` or some other initial calls...

**Comment by mathetake** [comment]
Anyway, I don't think it's the premature optimization. We would better figure out in this PR.

**Comment by PiotrSikora** [comment]
> > One thing that we need to consider, though, is that now those strings will be always evaluated, and calls to Envoy will be always made, even when the log level is other than trace, which results in wasted compute.
> 
> yeah, that's exactly what I am thinking about. strace should be triggered by "userland" so how about having ABI to enable/disable strace in host?

Erm, no, I meant that if Envoy starts with log level `trace` then we should log the ABI calls, but when it starts with lower log level, then we shouldn't even need to construct the string and evaluate parameter values.

I definitely don't think that it should be controlled by the Proxy-Wasm modules (tracing calls is used to observe them, not the other way around), so we don't need the ABI support for that.

**Comment by mathetake** [comment]
right. OK then I will add the support for getting log level from Envoy

**Comment by mathetake** [comment]
expected change in Envoy side: https://github.com/envoyproxy/envoy/pull/14433/files

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Pull request merged**